lib_eio/unix/thread_pool.ml

```diff
@@ -0,0 +1,216 @@
+   on stand-by, without having to wait for them to shut down automatically.
+   Without this termination mechanism, domains that have executed a call on
+   a systhread in the last [max_wait_seconds] seconds would have to wait for
+   their timeout to occur, introducing long unwanted pauses. *)
+let terminate pool =
+  pool.terminating <- true;
+  if not (Bitfield.is_empty pool.available) then
+    for i = 0 to max_standby_systhreads_per_domain - 1 do
+      if Bitfield.get pool.available i
+      then Mailbox.put pool.threads.(i) Exit
+    done
+
+(* [@poll error] ensures this function is atomic within systhreads of a single domain.
+   This function (re-)adds the worker systhread to the pool of available threads,
+   or exits if the pool is already at maximum capacity. *)
+let[@poll error] keep_thread_or_exit pool (mbox : Mailbox.t) =
+  if pool.terminating || Bitfield.is_full pool.available
+  then raise Thread.Exit
+  else (
+    let i = Bitfield.find_empty pool.available in
+    mbox.i <- i;
+    Bitfield.set pool.available i true;
+    pool.threads.(i) <- mbox
+  )
+
+(* [@poll error] ensures this function is atomic within systhreads of a single domain.
+   This function tries to reserve one available thread from the pool.
+   Since we cannot return an option in [@poll error] code, we simulate
+   an option by conditionally updating the reference [res].
+   The returned boolean indicates whether we updated [res] or not. *)
+let[@poll error] try_get_thread (pool : Mailbox.t t) res =
+  if Bitfield.is_empty pool.available
+  then false
+  else (
+    let i = Bitfield.find_ready pool.available in
+    Bitfield.set pool.available i false;
+    res := pool.threads.(i);
+    true)
+
+let make_thread pool =
+  let mbox = Mailbox.create () in
+  let _t : Thread.t = Thread.create (fun () ->
+    while true do
+      match Mailbox.take pool mbox with
+      | New -> assert false
+      | Exit -> raise Thread.Exit
+      | Job { fn; enqueue } ->
+        enqueue (try Ok (fn ()) with exn -> Error exn);
+        keep_thread_or_exit pool mbox
+    done
+  ) ()
+  in
+  mbox
+
+(* https://v2.ocaml.org/manual/parallelism.html#s:par_systhread_interaction
+   "Only one systhread at a time is allowed to run OCaml code on a particular domain."
+   So we keep a separate threadpool per domain. *)
+let key = Domain.DLS.new_key create
+
+(* [@poll error] ensures this function is atomic within systhreads of a single domain.
+   https://github.com/ocaml/ocaml/pull/12724
+   As of OCaml 5.1, [Domain.at_exit] is still not threadsafe *)
+let[@poll error] needs_init pool =
+  if pool.initialized
+  then false
+  else (pool.initialized <- true; true)
+
+let run_on_systhread ~enqueue fn =
+  let pool = Domain.DLS.get key in
+  if needs_init pool then Domain.at_exit (fun () -> terminate pool);
+  let mbox =
+    let res = ref Mailbox.dummy in
+    if try_get_thread pool res
+    then !res
+    else make_thread pool
```

Comment:
Hmm... Another way to organize the pool would be to separate the management of the threads from the queue of jobs to perform.  Basically, jobs would be put into a queue and worker threads would take jobs from the queue to perform.  This way you could avoid spawning a new thread for each new job after the pool capacity has been exhausted.

Reply:
Thank you for your review! I always appreciate you sharing your expertise in this area.

> This way you could avoid spawning a new thread for each new job after the pool capacity has been exhausted.

At the moment that's done on purpose. I absolutely did not want to risk user code deadlocking _only_ when the threadpool is at maximum capacity. This tends to cause bugs in production that are hard to discover during development. Do you have a recommendation?

> Basically, jobs would be put into a queue and worker threads would take jobs from the queue to perform.

I wanted to make sure that jobs would _always_ go to Thread 0 if Thread 0 is available, otherwise _always_ go to Thread 1 if Thread 1 is available, etc. This way we can guarantee that only the minimum number of threads are running at any given time because it allows the other threads to timeout and shutdown. In other words, the scheduling unfairness is on purpose. If there's a better way to achieve this effect, I'm all ears!

> Instead of having a separate mailbox (and semaphore) for every worker thread

In yesterday's dev meeting we discussed how @talex5 will be implementing the wait/timeout logic in the scheduler itself so we should be able to get rid of all the semaphores and all the C code I ported from my own [ocaml/ocaml#12867](https://github.com/ocaml/ocaml/pull/12867) PR. I do like your suggestion a lot though and may switch to it if beneficial after the scheduler changes.

Reply:
>  I absolutely did not want to risk user code deadlocking only when the threadpool is at maximum capacity.

That is a good point.

I think there are two completely different possible designs/semantics:
1. One that fundamentally has a hard limit on the number of systhreads to use to run things concurrently.
2. One that fundamentally has no such limit.

The current implementation is a bit of both.  It uses a fixed size array to manage the systhreads.  But, then, in case more concurrency is requested, it runs even more systhreads.

I think this adds to the complexity.  I suspect that changing the design to clearly implement one or the other could simplify the internal implementation.  Perhaps not (only) in terms of lines of code, but also in terms of conceptual clarity.  Indeed, it wasn't entirely clear to me that spawning more threads was really intentional.

So, I would recommend trying to rethink the internals with the idea that there is no limit on the number of systhreads.

Reply:
> This way we can guarantee that only the minimum number of threads are running at any given time because it allows the other threads to timeout and shutdown. In other words, the scheduling unfairness is on purpose. If there's a better way to achieve this effect, I'm all ears!

I see.  So, what you really want is to use the thread that was the most recent thread to become idle.  That is because then the threads that became idle earlier can timeout.   IOW, for picking the next idle thread to pass a new job to you want LIFO and for picking the next idle thread to die/kill you want FIFO.  Basically a deque, except that the idea here is that threads remove themselves from the deque.

I suspect a dynamic doubly-linked list could be used for this and it would likely be a bit simpler (fewer lines of code and also a simpler concept to understand) than the array with a bitfield.

But this is also assuming that threads timeout on their own.  An alternative approach could be to manage the threads externally.  Idle threads would add themselves to a data structure (along with a timestamp, for example).  Then some external mechanism would remove old threads.  With Eio this external mechanism could be something like a per domain fiber that periodically explicitly kills old idle threads when there are some idle threads.  This approach could also remove the need for timedwait (which I think really would be a great addition to the Stdlib, but it may take quite some time before we get that).